forage-android/src/main/java/com/joinforage/forage/android/core/element/state/PanElementStateManager.kt

```diff
@@ -74,31 +69,84 @@ class PanElementStateManager(state: ElementState) : ElementStateManager(state) {
-        }
+open class WhitelistedCards(
+    private val prefix: String,
+    private val repeatCount: Int = 1
+) : PanValidator {
+    override fun checkIfValid(cardNumber: String): Boolean {
+        // for example 444444 4444 4444 *** **
+        // maps to 44444444444444*****, which could be
+        // a whitelisted pattern
+        return cardNumber.startsWith(prefix.repeat(repeatCount))
+    }
+
+    override fun checkIfComplete(cardNumber: String): Boolean {
+        return checkIfValid(cardNumber) && cardNumber.length in MIN_CARD_LENGTH..MAX_CARD_LENGTH
     }
 
-    private fun setIsEmpty(cardNumber: String) {
-        this.isEmpty = cardNumber.isEmpty()
+    override fun checkForValidationError(cardNumber: String): ElementValidationError? {
+        // this is a developer-only validator and should remain
+        // quiet by not returning a validationError
+        return null
     }
+}
+
+class PaymentCaptureErrorCard : WhitelistedCards("4", 14)
+class BalanceCheckErrorCard : WhitelistedCards("5", 14)
+class NonProdValidEbtCard : WhitelistedCards("9", 4)
+class EmptyEbtCashBalanceCard : WhitelistedCards("654321")
 
-    fun handleChangeEvent(newCardNumber: String) {
-        setIsValid(newCardNumber)
-        setValidationError(newCardNumber)
-        setIsComplete(newCardNumber)
-        setIsEmpty(newCardNumber)
+class PanElementStateManager(state: ElementState, private val validators: Array<PanValidator>) : ElementStateManager(state) {
+
+    private fun checkIsValid(cardNumber: String): Boolean {
+        return validators.any { it.checkIfValid(cardNumber) }
+    }
+    private fun checkIfComplete(cardNumber: String): Boolean {
+        return validators.any { it.checkIfComplete(cardNumber) }
+    }
+    private fun checkForValidationError(cardNumber: String): ElementValidationError? {
+        return validators
+            .map { it.checkForValidationError(cardNumber) }
+            .firstOrNull { it != null }
+    }
+
+    fun handleChangeEvent(rawInput: String) {
+        // because the input may be formatted, we need to
+        // strip everything but digits
+        val newCardNumber = rawInput.filter { it.isDigit() }
+
+        // check to see if any of the validators believe the
+        // card to be valid
+        this.isValid = checkIsValid(newCardNumber)
+        this.isComplete = checkIfComplete(newCardNumber)
+        this.validationError = checkForValidationError(newCardNumber)
+        this.isEmpty = newCardNumber.isEmpty()
+
+        // invoke the registered listener with the updated state
         onChangeEventListener?.invoke(getState())
     }
 
     companion object {
         fun forEmptyInput(): PanElementStateManager {
-            return PanElementStateManager(INITIAL_ELEMENT_STATE)
+            return PanElementStateManager(
+                INITIAL_ELEMENT_STATE,
+                arrayOf(StrictEbtValidator())
+            )
+        }
+
+        fun DEV_ONLY_forEmptyInput(): PanElementStateManager {
```

Comment:
I see this being used in tests as well as a check to use this whenever we aren't in `prod`. If this is entirely for tests, leaving it as is makes sense. If this is the function that should be run if we aren't in production, could we change the name? I'm not sure what is best, but `NON_PROD_forEmptyInput`? What do you think?

Reply:
No issue with changing the name to `NON_PROD_forEmptyInput`! And, apologies, I don't think I fully grasped your whole remark but `NON_PROD_forEmptyInput` is tested in tests, but its main usage is in `ForagePANEditText` [when instantiating the `manager` property](https://github.com/teamforage/forage-android-sdk/pull/69/files/554202fe522df990691dc583e5d970ca5719e8f8#diff-8c917c681bf8784c25ec13dd769ed68175e6e84897fdcc64dbb20c3ddab7ee5eR38-R44). Reproducing here for convenience:

```kotlin
 private val manager: PanElementStateManager = if (BuildConfig.FLAVOR == "prod") {
        // strictly support only valid Ebt PAN numbers
        PanElementStateManager.forEmptyInput()
    } else {
        // allows whitelist of special Ebt PAN numbers
        PanElementStateManager.DEV_ONLY_forEmptyInput() // will be renamed to NON_PROD_forEmptyInput()
    }
```

Reply:
Addressed in #72 to make it easier to see changes!